ZOOKEEPER-3172: Quorum TLS - fix port unification to allow rolling upgrades

ivmaykov opened the pull request:
Fix numerous problems with UnifiedServerSocket, such as hanging the accept() thread when the client doesn't send any data or crashing if less than 5 bytes are read from the socket in the initial read.

Re-enable the "portUnification" config option.

## Fixed networking issues/bugs in UnifiedServerSocket

- don't crash the `accept()` thread if the client closes the connection without sending any data
- don't corrupt the connection if the client sends fewer than 5 bytes for the initial read
- delay the detection of TLS vs. plaintext mode until a socket stream is read from or written to. This prevents the `accept()` thread from getting blocked on a `read()` operation from the newly connected socket.
- prepending 5 bytes to `PrependableSocket` and then trying to read >5 bytes would only return the first 5 bytes, even if more bytes were available. This is fixed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2501/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2522/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2529/

ivmaykov commented:
I should mention that this code has been internally reviewed at Facebook, has been landed on our internal fork, and has been running in production for weeks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2533/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2543/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2547/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2556/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2560/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2564/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2585/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2592/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2594/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2598/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build/2600/

ivmaykov commented:
Rebase on latest master, no longer includes #678 as it has been merged upstream.

anmolnar commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/UnifiedServerSocketTest.java:
```diff
@@ -17,156 +17,644 @@
+    private static final class UnifiedServerThread extends Thread {
+        private final byte[] dataToClient;
+        private List<byte[]> dataFromClients;
+        private List<Thread> workerThreads;
```
Use some `Executor` instead?

ivmaykov replied:
I used explicit threads because that's what we do in Leader.java - create a dedicated Thread per connection. I wanted the threading in the test to be as close as possible to the real use case. I could change it to an Executor if you want, but don't see that much value in such a change.

anmolnar replied:
Creating tests as close as possible to the original impl is nice, but you're still testing a different thing anyway. Executors are good for dealing with threads and I think they're preferred over manual thread creation if acceptable. There're various different kinds of them: you can control the thread creation behaviour, number of threads, etc. And I think it leads to more readable code.

ivmaykov replied:
Sure, will change it to an Executor

ivmaykov replied:
Done

anmolnar commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/UnifiedServerSocketTest.java:
```diff
@@ -17,156 +17,644 @@
-        handshakeCompleted = false;
-
         port = PortAssignment.unique();
+        localServerAddress = new InetSocketAddress("localhost", port);
```
Using `InetAddress.getLoopbackAddress()` ?

ivmaykov replied:
Sure, I could do that if you prefer.

anmolnar replied:
Thanks.

ivmaykov replied:
Done

anmolnar commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/UnifiedServerSocketTest.java:
```diff
@@ -17,156 +17,644 @@
+     */
+    @Test
+    public void testDenialOfServiceResistanceNonStrictServer() throws Exception {
+        UnifiedServerThread serverThread = new UnifiedServerThread(
```
In these DOS tests you're actually testing the `UnifiedServerThread` implementation which is a test-only thing. Does it make sense or am I missing something here?

In "reality" the accept thread is basically the `Leader` and the one which does TLS handshake is the `LearnerHandler`.

ivmaykov replied:
This is true. I tried to make the threading behavior in this test as similar as possible to `Leader`. I don't think Leader should be involved in this unit test, but perhaps a similar "denial of service resistance" test case can be added to some quorum test. Do you know which unit test would be a good place for such a test case?

anmolnar replied:
I would create new test class inherited from `QuorumPeerTestBase` dedicated to DOS tests, but I think this topic leads to another: stress testing ZooKeeper. I highly support the effort, but please remove them from this PR and create separate Jira for that.

ivmaykov replied:
How about we keep the unit test that's focused strictly on UnifiedServerSocket in this PR, and I'll open a separate PR/Jira for adding a new unit test that tests `Leader`? There is some value in the existing tests, since they did catch issues with the original UnifiedServerSocket implementation.

anmolnar replied:
@ivmaykov That's pretty much the same as I suggested: keep all unit tests except the ones which contains the word "dos". Also please open a new jira about adding DOS tests.

ivmaykov commented on zookeeper-server/src/main/java/org/apache/zookeeper/common/X509Util.java:
```diff
@@ -350,14 +389,22 @@ public static X509TrustManager createTrustManager(
+        }
         configureSSLSocket(sslSocket);
-
+        sslSocket.setUseClientMode(false);
```
Yes and yes. In #681 I make the client auth setting configurable.

ivmaykov commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/PrependableSocket.java:
```diff
@@ -18,32 +18,47 @@
 public class PrependableSocket extends Socket {
 
-  private SequenceInputStream sequenceInputStream;
+  private PushbackInputStream pushbackInputStream;
```
I don't know why it worked like that, but that is what I observed in tests. It looks like SequenceInputStream does not join returned data across boundaries of the underlying streams when the first stream gets to EOF. I don't think this is desired behavior since it would cause the stream to return 5 bytes when more than 5 bytes are actually available.

ivmaykov replied:
Yes, SequenceInputStream likely still works in practice, but exposes the discontinuity between the underlying streams to the caller. Using PushbackInputStream hides it and I think is preferable.

ivmaykov commented:
@anmolnar use Executor and InetAddress.getLoopbackAddress()

ivmaykov commented:
@anmolnar the benefit of the current DOS tests is to verify that certain actions on a UnifiedSocket (getting input stream, setting socket options, etc) do not trigger the blocking read for TLS / plaintext mode detection. I think there is benefit to having such tests `UnifiedServerSocketTest`, since they test behavior that's implemented in `UnifiedServerSocket.java`. What do you think?

anmolnar commented:
@ivmaykov I did another cycle of review the unit tests, sorry I still not see value in denial-of-service tests, but maybe I don't see something important. Let's put it this way:

Use case of `UnifiedServerSocket` is the following:
> Extend standard Socket class to catch all read events in order to trigger TLS/Plaintext mode detection in a lazy fashion (catching the last chance of doing so).

In order to do that it captures calls to the underlying input/output streams and initiates detection of channel type on the first read/write operation. So far so good.

The important bit here is that `UnifiedServerSocket` **doesn't contain anything related to threading**. As a consequence if we write purely unit tests for this class we won't have to verify anything which is related to threading.

DOS test comment says:
> This test makes sure that UnifiedServerSocket used properly (a single thread accept()-ing connections and handing the resulting sockets to other threads for processing) is not vulnerable to a simple denial-of-service attack in which a client connects and never writes any bytes. **This should not block the accepting thread, since the read to determine if the client is sending a TLS handshake or not happens in the processing thread.**

And here's the thing: this test is testing the proper implementation of the server (e.g. dealing with threads in the right way), which is implemented in the test itself: `UnifiedServerThread`.

My 2 cents here is that if you think you haven't covered a user scenario or an edge with the existing tests, you need to rewrite these tests to be more strict and test that-and-only-that particular case which is missing.

I think the coverage is acceptable (what you mentioned in your latest comment has already been covered), but there's no harm in adding more. I just don't want tests which don't add value **or** adds value in an unreasonably cumbersome way.

Please correct me if I'm wrong.

ivmaykov commented:
@anmolnar I understand your point about the DoS tests, and agree that they are not very focused and do end up testing the structure of the test code itself in (i.e. that the implementation of `UnifiedServerSocketTest$UnifiedServerThread` plays nicely with `UnifiedServerSocket`). Let me explain really briefly how that test came about, since the history behind it is missing in this discussion.

Initially when I rewrote `UnifiedServerSocket` and added the `UnifiedSocket` inner class, I didn't have the `UnifiedInputStream` and `UnifiedOutputStream` inner classes. I was careful to avoid any reads inside `accept()`, and thought that was good enough. However, with this version of the code we still saw problems on our test cluster during a stress disconnect/reconnect test. Investigation showed evidence that the Leader's `accept()` thread was getting stuck, and then I found the code in `Leader$LearnerCnxAcceptor#run()` that calls `getInputStream()` on the accepted socket and wraps the result in a `BufferedInputStream`. The only way at the time to get the input stream was to resolve the type of socket (by doing a read of first 5 bytes and then TLS/plaintext mode detection), and get the real socket's input stream.

My fix involved adding the `UnifiedInputStream` and `UnifiedOutputStream` inner classes, which allowed me to delay the mode detection. Now, instead of triggering mode detection at the time of `getInputStream()`, we can trigger it the first time the stream is used for I/O. I wrote the DoS test as part of the same diff, and was careful to replicate the threading behavior of `Leader$LearnerCnxAcceptor.run()` in `UnifiedServerThread`. I also verified that the new tests failed on a version of the code that didn't have the unified stream wrappers.

If I remove the DoS tests, the behavior of delaying the mode detection's read to the point of first I/O operation would not be tested properly. I think that code should be tested! However, if you prefer I could probably test it in a more focused way (i.e. just test each method that's meant to be non-blocking by itself). Would that be an acceptable compromise?

ivmaykov commented:
@anmolnar removed DoS tests from `UnifiedServerSocketTest.java`. Added a new unit test `UnifiedServerSocketModeDetectionTest.java` which checks that calling certain methods on the server side of a unified socket connection does not trigger mode detection. Also rebased on latest master to fix merge conflict.

anmolnar commented:
@ivmaykov Gosh! I just did another round of review with your tests today and was tending to accept the DoS tests with just renaming and amending the comment. :)
Sorry for not responding earlier, I'll check your new code tomorrow and give feedback.

ivmaykov commented:
I like the old DoS tests, I'm happy to restore them. The new unit test I added can stay too.

anmolnar commented:
@ivmaykov I'm not a fan of them, but understand the intention and can't figure out a more straightforward way to test the behaviour. So let's go with them.

What do you think of renaming them to something like `testTLSDetectionNonBlocking(Non)StrictServer`?
I would also rephrase the comment to emphasize that the tests are intended to validate that TLS detection is non-blocking, so it's safe to getInputStream() on the accept thread.

These are not really denial-of-service tests IMHO. Thoughts?

ivmaykov commented:
@anmolnar renamed the tests and cleaned them up quite a bit too (the formerly-DoS tests don't actually need a background thread for the bad client socket, which simplifies things).

anmolnar commented:
Thanks @ivmaykov !
Committed to master branch.
Please create separate PR for 3.5, because it has some merge conflicts.
